src/mapgen.cpp

```diff
@@ -193,44 +194,153 @@ void mapgen_function_builtin::generate( mapgendata &mgd )
 ///// mapgen_function class.
 ///// all sorts of ways to apply our hellish reality to a grid-o-squares
 
+class mapgen_basic_container
+{
+    private:
+        std::vector<std::shared_ptr<mapgen_function>> mapgens_;
+        weighted_int_list<std::shared_ptr<mapgen_function>> weights_;
+
+    public:
+        int add( const std::shared_ptr<mapgen_function> ptr ) {
+            assert( ptr );
+            mapgens_.push_back( ptr );
+            return mapgens_.size() - 1;
+        }
+        void erase( const size_t index ) {
+            if( index > mapgens_.size() ) {
+                return;
+            }
+            assert( mapgens_[index] );
+            // Can't actually erase it as this might invalid the index stored elsewhere
+            mapgens_[index]->weight = 0;
+        }
+        /**
+         * Pick a mapgen function randomly.
+         * @p hardcoded_weight Weight for an additional entry. If that entry is chosen,
+         * a null pointer is returned. If unsure, just use 0 for it.
+         */
+        mapgen_function *pick( const int hardcoded_weight ) const {
+            if( hardcoded_weight > 0 &&
+                rng( 1, weights_.get_weight() + hardcoded_weight ) > weights_.get_weight() ) {
+                return nullptr;
+            }
+            const std::shared_ptr<mapgen_function> *const ptr = weights_.pick();
+            if( !ptr ) {
+                return nullptr;
+            }
+            assert( *ptr );
+            return ptr->get();
+        }
+        /**
+         * Calls @ref mapgen_function::setup and sets up the internal weighted list using
+         * the **current** value of @ref mapgen_function::weight. This value may have
+         * changed since it was first added, so this is needed to recalculate the weighted list.
```

Comment:
Could you please elaborate on this. Why would `weight` change after adding to the container? Maybe it should take ownership of the `mapgen_function`'s themselves (instead of possessing mere pointers) and protect its own guts form external interference? The current scheme seems to be rather fragile.

Reply:
> Why would weight change after adding to the container?

When `erase` at line 209 is called.

Reply:
> instead of possessing mere pointer

Nah, pointers are exposed e.g. via `load_mapgen_function`. If I read the code correctly, this can be used to connect one mapgen instance with multiple overmap terrains (equivalent to giving each overmap terrain a copy of the mapgen). It seems like the mapgen in "lab/lab_trains.json" make use of that.

Reply:
I see, thanks. It would be ideal to refrain form `int` indexes at some point and use persistent iterators instead. Resetting `weight` to `0` in a temporary container as a sign of an item being `erased`, and then not being able to operate on `indexes` after invocation of `setup` looks, well, a bit hackish.

If you still decide to stick to this approach, it would be nice to issue a `debugmsg` in case someone tries to call `add` or `erase` after `setup`.